src/libmongoc/src/mongoc/mongoc-ts-pool.c

```diff
@@ -0,0 +1,263 @@
+ */
+static bool
+_should_prune (const pool_node *node)
+{
+   mongoc_ts_pool *pool = node->owner_pool;
+   return pool->params.prune_predicate &&
+          pool->params.prune_predicate (node->data, pool->params.userdata);
+}
+
+/**
+ * @brief Create a new pool node and contained element.
+ *
+ * @return pool_node* A pointer to a constructed element, or NULL if the
+ * constructor sets `error`
+ */
+static pool_node *
+_new_item (mongoc_ts_pool *pool, bson_error_t *error)
+{
+   pool_node *node =
+      bson_malloc0 (sizeof (pool_node) + pool->params.element_size);
+   node->owner_pool = pool;
+   if (pool->params.constructor) {
+      /* To construct, we need to know if that constructor fails */
+      bson_error_t my_error;
+      if (!error) {
+         /* Caller doesn't care about the error, but we care in case the
+          * constructor might fail */
+         error = &my_error;
+      }
+      /* Clear the error */
+      error->code = 0;
+      error->domain = 0;
+      error->message[0] = 0;
+      /* Construct the object */
+      pool->params.constructor (node->data, pool->params.userdata, error);
+      if (error->code != 0) {
+         /* Constructor reported an error. Deallocate and drop the node. */
+         bson_free (node);
+         node = NULL;
+      }
+   }
+   if (node) {
+      bson_atomic_int_add (&pool->outstanding_items, 1);
+   }
+   return node;
+}
+
+/**
+ * @brief Destroy the given node and the element that it contains
+ */
+void
+_delete_item (pool_node *node)
+{
+   mongoc_ts_pool *pool = node->owner_pool;
+   if (pool->params.destructor) {
+      pool->params.destructor (node->data, pool->params.userdata);
+   }
+   bson_free (node);
+}
+
+/**
+ * @brief Try to take a node from the pool. Returns `NULL` if the pool is empty.
+ */
+static pool_node *
+_try_get (mongoc_ts_pool *pool)
+{
+   pool_node *node;
+   bson_mutex_lock (&pool->mtx);
+   node = pool->head;
+   if (node) {
+      pool->head = node->next;
+   }
+   bson_mutex_unlock (&pool->mtx);
+   if (node) {
+      bson_atomic_int_add (&pool->size, -1);
+      bson_atomic_int_add (&pool->outstanding_items, 1);
+   }
+   return node;
+}
+
+mongoc_ts_pool *
+mongoc_ts_pool_new (mongoc_ts_pool_params params)
+{
+   mongoc_ts_pool *r = bson_malloc0 (sizeof (mongoc_ts_pool));
+   r->params = params;
+   r->head = NULL;
+   r->size = 0;
+   r->outstanding_items = 0;
+   bson_mutex_init (&r->mtx);
+   return r;
+}
+
+void
+mongoc_ts_pool_free (mongoc_ts_pool *pool)
+{
+   BSON_ASSERT (pool->outstanding_items == 0 &&
+                "Pool was destroyed while there are still items checked out");
+   mongoc_ts_pool_clear (pool);
+   bson_mutex_destroy (&pool->mtx);
+   bson_free (pool);
+}
+
+void
+mongoc_ts_pool_clear (mongoc_ts_pool *pool)
+{
+   pool_node *node;
+   {
+      bson_mutex_lock (&pool->mtx);
+      node = pool->head;
+      pool->head = NULL;
+      pool->size = 0;
+      bson_mutex_unlock (&pool->mtx);
+   }
+   while (node) {
```

Comment:
Would it be possible (or reasonable) to redefine this in terms of `visit_each`?

Reply:
Possible: yes, but not desirable, since visit_each holds a lock while it visits every element. This quickly clears the pool and then unlocks it before it starts deleting elements.